python/cuml/decomposition/tsvd_mg.pyx

```diff
@@ -23,570 +22,548 @@ import ctypes
 
-        self.noise_variance_ = cudf.Series(zeros(1, dtype=self.dtype))
+class TSVDMG(TruncatedSVD):
 
-    def _get_ctype_ptr(self, obj):
-        # The manner to access the pointers in the gdf's might change, so
-        # encapsulating access in the following 3 methods. They might also be
-        # part of future gdf versions.
-        return obj.device_ctypes_pointer.value
+    def __init__(self, **kwargs):
+        super(TSVDMG, self).__init__(**kwargs)
 
-    def _get_column_ptr(self, obj):
-        return self._get_ctype_ptr(obj._column._data.to_gpu_array())
-
-    def _get_gdf_as_matrix_ptr(self, gdf):
-        return self._get_ctype_ptr(gdf.as_gpu_matrix())
-
-    def _fit_spmg(self, X, _transform=True, gpu_ids=[]):
-
-        if (not isinstance(X, np.ndarray)):
-            msg = "X matrix must be a Numpy ndarray. Dask will be supported" \
-                  + " in the next version."
-            raise TypeError(msg)
-
-        n_gpus = len(gpu_ids)
-
-        n_rows = X.shape[0]
-        n_cols = X.shape[1]
-        self.params.n_rows = n_rows
-        self.params.n_cols = n_cols
+    def _initialize_arrays(self, n_components, n_rows, n_cols):
 
-        if (n_rows < 2):
-            msg = "X must have at least two rows."
-            raise TypeError(msg)
+        self.components_ary = rmm.to_device(zeros(n_components*n_cols,
+                                                  dtype=self.dtype))
+
+        self.explained_variance_ = cudf.Series(zeros(n_components,
+                                               dtype=self.dtype))
+
+        self.explained_variance_ratio_ = cudf.Series(zeros(n_components,
+                                                     dtype=self.dtype))
+
+        self.singular_values_ = cudf.Series(zeros(n_components,
+                                                  dtype=self.dtype))
+
+    def _build_dataFloat(self, arr_interfaces):
+        cdef floatData_t ** dataF = < floatData_t ** > \
+            malloc(sizeof(floatData_t *)
+                   * len(arr_interfaces))
+
+        cdef uintptr_t input_ptr
+        for x_i in range(len(arr_interfaces)):
+            x = arr_interfaces[x_i]
+            input_ptr = x["data"]
+            dataF[x_i] = < floatData_t * > malloc(sizeof(floatData_t))
+            dataF[x_i].ptr = < float * > input_ptr
+            dataF[x_i].totalSize = < size_t > x["shape"][0]
+        return <size_t>dataF
+
+    def _build_dataDouble(self, arr_interfaces):
+        cdef doubleData_t ** dataD = < doubleData_t ** > \
+            malloc(sizeof(doubleData_t *)
+                   * len(arr_interfaces))
+
+        cdef uintptr_t input_ptr
+        for x_i in range(len(arr_interfaces)):
+            x = arr_interfaces[x_i]
+            input_ptr = x["data"]
+            dataD[x_i] = < doubleData_t * > malloc(sizeof(doubleData_t))
+            dataD[x_i].ptr = < double * > input_ptr
+            dataD[x_i].totalSize = < size_t > x["shape"][0]
+        return <size_t>dataD
```

Comment:
Many of this methods seem like would be appropriate to be shared among different classes, could you create a gh issue to look into this afterwards?

Reply:
Created this PR https://github.com/rapidsai/cuml/issues/1400. Thanks!